Allow scheduling a (example) product via the web UI

Martchus opened the pull request:
Related ticket: https://progress.opensuse.org/issues/166658

---

~~Still a draft because I need to add UI tests and move the so far hard-coded data into `openqa.ini`. It already reads scenario definitions and certain parameters directly from the example distribution, though.~~ Of course this only works if there's a checkout of the example distribution. Maybe we should offer the user to create that checkout in case it is missing (which wouldn't be very hard to do; we'd just add a link on the web UI that'll internally spawn a checkout Minion job we recently implemented anyway).

---

This is how the menu, the form and success/error messages look like in one screenshot:

![grafik](https://github.com/user-attachments/assets/06959fd5-8e7d-4be3-9760-143c26ff88d5)

okurz commented:
overall this looks quite good. I was thinking of alternatives where to place the action but your proposal so far looks like the best to me. Just with the "+" I was thinking of a bigger button, possibly blue, with a frame around to distinguish clearly from the caret

Martchus commented:
I'm pondering about configurability. Reading the structure that has currently the FIXME comment from `openqa.ini` doesn't really work because it needs nesting but we only use INI. I suppose I could simply read those values from an additional file (e.g. `test_presets.ini`). Otherwise I could check for sections like `[test_presets/example]` in `openqa.ini` but it'll be an additional loop.

Martchus commented:
I now made it configurable. So in this screenshot I overrode the title by putting a `[test_presets/example]` section to my `openqa.ini`:

![screenshot_20240918_143341](https://github.com/user-attachments/assets/7d501c39-63a7-42bb-849d-9e1b49202bfe)

This screenshot also shows that it is now also possible to add additional settings. I think this is important if we want to offer this as a generic form (and the implementation is also quite simple).

This leaves tests and the possibility to clone the example distribution automatically (for the scenario definitions).

perlpunk commented:
> I'm pondering about configurability. Reading the structure that has currently the FIXME comment from `openqa.ini` doesn't really work because it needs nesting but we only use INI. I suppose I could simply read those values from an additional file (e.g. `test_presets.ini`). Otherwise I could check for sections like `[test_presets/example]` in `openqa.ini` but it'll be an additional loop.
> 
> EDIT: I guess I'll go for `[test_presets/example]` because I suppose the additional loop can be avoided.

Config::IniFiles allows groups. You separate the group from the section with a space.
https://metacpan.org/pod/Config::IniFiles#Groups

Martchus commented:
> Config::IniFiles allows groups. You separate the group from the section with a space.
> https://metacpan.org/pod/Config::IniFiles#Groups

I now changed it to use e.g. `[test_preset foo]`. This way we could use the grouping mechanism of the INI parser - even though this is not necessary for the PR in its current form. With this we could use https://metacpan.org/pod/Config::IniFiles#GroupMembers-($group) to show the list of available presets in e.g. the navbar menu. (Not implemented yet. For now I stick to what we have planned in the ticket. So cloning the example repo automatically is probably more important.)

kalikiana commented:
> Still a draft because I need to add UI tests and move the so far hard-coded data into openqa.ini. It already reads scenario definitions and certain parameters directly from the example distribution, though. Of course this only works if there's a checkout of the example distribution. Maybe we should offer the user to create that checkout in case it is missing (which wouldn't be very hard to do; we'd just add a link on the web UI that'll internally spawn a checkout Minion job we recently implemented anyway).

Is this obsolete? As this is not a draft anymore 😉

kalikiana reviewed (approved):
Looks really nice. Especially with the help texts. And I appreciate your making it extensible in the future.

Martchus commented:
I crossed it the obsolete part.

perlpunk commented on lib/OpenQA/WebAPI/Controller/Test.pm:
```diff
@@ -116,6 +117,51 @@ sub list {
+    return undef unless my $ini_config = $config->{ini_config};
+    my $ini_key = "test_preset $preset_key";
+    return $presets{$preset_key}
+      = $ini_config->SectionExists($ini_key)
```
I still don't like that we are creating some extra code for this specific feature to handle nested data, while we could do this in a more generic way in Setup.pm with the group feature.
Also it moves any validation we might want to do to runtime instead of server startup.
Since it's something internal we can probably change the implementation later, but just saying.
Not sure what others think.

okurz commented:
> I crossed out the obsolete part.
> 
> Considering there's remaining work to do I'm still considering this as not ready. We don't want to show a new option prominently that'll not work out of the box.

Consider a feature switch which should be off by default until we stabilized the feature

Martchus commented:
I now implemented cloning the example test distribution. So when the user enters the page and the repo hasn't been cloned yet it says "You have to **clone the example test distribution** first." where the bold part is a link, When clicking on the link one gets a loading indication and after a few seconds (it really doesn't take long to checkout the example distribution) the form reloads with all values filled in. (And in any error case the error is shown and one can retry.) This didn't require adding much code because we already have Minion tasks for cloning from Git now.

perlpunk commented:
Btw, we have to be careful to not do this on o3 or osd until fetchneedles is replaced. The example distri has a default branch `main`, and fetchneedles can't handle different default branches.

okurz commented on etc/openqa/openqa.ini:
```diff
@@ -325,3 +325,8 @@ concurrent = 0
 # home =
 # username = openqa-user
 # ssh_key_file = ~/.ssh/id_rsa
+
+[test_preset bar]
+title = Some preset
+distri = does-not-exist
+casedir = http://foo.git
```
Those should be commented by default. And can you add some explanation comments here? I wouldn't know why I would want to set those settings. And please use the actual default values

Martchus replied:
This should have been added to the config file we use for tests. I didn't intend to touch the normal config template at all for now.

okurz commented on lib/OpenQA/WebAPI/Controller/Test.pm:
```diff
@@ -116,6 +117,71 @@ sub list {
+sub create ($self) {
+    my $preset_key = $self->param('preset');
+    my $preset = $self->_load_test_preset($preset_key);
+    if (defined $preset) {
+        $self->stash(flash_info => $preset->{info});
+        $self->_load_scenario_definitions($preset);
+    }
+    elsif (defined $preset_key) {
+        $self->stash(flash_error => "The specified preset '$preset_key' does not exist.");
+    }
```
Can this be a

```suggestion
    return $self->stash(flash_error => "The specified preset '$preset_key' does not exist.") if defined $preset_key;
    if (defined $preset) {
        $self->stash(flash_info => $preset->{info});
        $self->_load_scenario_definitions($preset);
    }
```

although I wonder why should we error out if `$preset_key` is defined and not if it's *not* defined?

okurz commented on t/ui/29-create_tests.t:
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env perl
+# Copyright 2024 SUSE LLC
```
```suggestion
# Copyright SUSE LLC
```

okurz reviewed (changes requested):
Some inline comments.

And have you considered https://github.com/os-autoinst/openQA/pull/5933#issuecomment-2361040754 ?

Martchus commented:
> And have you considered https://github.com/os-autoinst/openQA/pull/5933#issuecomment-2361040754 ?

I only read that comment after I implemented the most recent changes I pushed to this PR yesterday. I don't think it is useful to merge this feature without enabling it by default because it is about the out of the box experience.

okurz commented:
> > And have you considered [#5933 (comment)](https://github.com/os-autoinst/openQA/pull/5933#issuecomment-2361040754) ?
> 
> I only read that comment after I implemented the most recent changes I pushed to this PR yesterday. I don't think it is useful to merge this feature without enabling it by default because it is about the out of the box experience.

I know but do you trust it enough to be enabled on all production instances? After all the primary goal was to show that easy starting point on *clean jobless* instances

Martchus commented:
> I know but do you trust it enough to be enabled on all production instances?

The feature itself I'd generally consider stable. It has full test coverage. The checkout of the example distribution relies on the Git cloning we implemented a while ago and I don't think it is generally problematic.

 I suppose the only problem is fetchneedles cannot handle the use of `main` as main branch (as opposed to `master`). At least according to @perlpunk. So I'll have a look at the situation. Maybe that's easy to fix or we can ignore it until fetchneedles is replaced or I can do a manual checkout on our production instances that uses and alias for master.

Martchus commented:
I added another commit to avoid problems with `fetchneedles`. See the commit message for details. The short story is that it'll work on our production instances and I also did local test runs of the entire script. There's only one caveat which I don't find problematic but I added a comment in the `fetchneedles` script documenting it.

kalikiana commented on script/fetchneedles:
```diff
@@ -72,24 +72,40 @@ deal_with_stale_lockfile() {
 
+# Determine the "main branch" for the remote origin unless a branch has been specified explicitly
+# note: This relies on a file called `refs/remotes/origin/HEAD` with contents like `ref: refs/remotes/origin/master` being
+#       present under the `.git` direcgtory. This should be the case for all recently cloned Git repositories. Otherwise
```
```suggestion
#       present under the `.git` directory. This should be the case for all recently cloned Git repositories. Otherwise
```

okurz commented on t/data/openqa.ini:
```diff
@@ -1 +0,0 @@
-../../etc/openqa/openqa.ini
```
why?

Martchus replied:
Because I had to add config values for testing instead of just testing with the default config. (We want to cover the part of the code that allows the users/admins to override values so we obviously need to supply this kind of configuration when running the test.)

perlpunk replied:
In other tests we just change the config on the fly to test specific features

okurz reviewed (approved):
Looks very nice now. Fix the code style issues and I am sure I can approve.

okurz reviewed (changes requested):
Looks very nice now. Fix the code style issues and I am sure I can approve.

okurz commented on lib/OpenQA/Setup.pm:
```diff
@@ -198,8 +198,15 @@ sub read_config ($app) {
-        carry_over => \%CARRY_OVER_DEFAULTS
-    );
+        carry_over => \%CARRY_OVER_DEFAULTS,
+        'test_preset example' => {
+            title => 'Create example test',
+            info => 'Parameters to create an example test have been pre-filled in the following form. '
+              . 'You can simply submit the form as-is to test your openQA setup.',
+            casedir => 'https://github.com/os-autoinst/os-autoinst-distri-example.git',
+            distri => 'example',
+            build => 'openqa',
+        });
```
Where did the according example changes to the openqa.ini file go? We should still have the according sections pre-filled and commented out for all possible options
